firedrake/slate/slac/kernel_builder.py

```diff
@@ -144,18 +144,18 @@ def collect_tsfc_kernel_data(self, mesh, tsfc_coefficients, tsfc_constants, wrap
 
         # Pick the coefficients associated with a Tensor()/TSFC kernel
         tsfc_coefficients = {tsfc_coefficients[i]: indices for i, indices in kinfo.coefficient_numbers}
-        for c, cinfo in wrapper_coefficients.items():
-            if c in tsfc_coefficients:
-                if isinstance(cinfo, tuple):
-                    if tsfc_coefficients[c]:
-                        ind, = tsfc_coefficients[c]
-                        if ind != 0:
-                            raise ValueError(f"Active indices of non-mixed function must be (0, ), not {tsfc_coefficients[c]}")
-                        kernel_data.append((c, cinfo[0]))
-                else:
-                    for ind, (c_, info) in enumerate(cinfo.items()):
-                        if ind in tsfc_coefficients[c]:
-                            kernel_data.append((c_, info[0]))
+        for c in tsfc_coefficients:
+            cinfo = wrapper_coefficients[c]
+            if isinstance(cinfo, tuple):
+                if tsfc_coefficients[c]:
+                    ind, = tsfc_coefficients[c]
+                    if ind != 0:
+                        raise ValueError(f"Active indices of non-mixed function must be (0, ), not {tsfc_coefficients[c]}")
+                    kernel_data.append((c, cinfo[0]))
+            else:
+                for ind, (c_, info) in enumerate(cinfo.items()):
+                    if ind in tsfc_coefficients[c]:
+                        kernel_data.append((c_, info[0]))
```

Comment:
If we need `kernel_data` to respect the ordering of `tsfc_coefficients`, then we could instead initialize it as a `dict.fromkeys(tsfc_coefficients)`, populate it, then pop keys where the value was left as None, and cast back as a list of tuples.
```
kernel_data = dict.fromkeys(tsfc_coefficients)

for ...
   if ...:
      kernel_data[c] = cinfo[0]
   else: 
      for ...
         if ...
   kernel_data[c_] = info[0]

kernel_data = list(kernel_data.items())
return kernel_data
```

Reply:
This only makes sense if we do not expect a coefficients appearing more than one time

Reply:
I think the way we currently do is more straightforward.